keps/sig-api-machinery/20190226-network-proxy.md

```diff
@@ -0,0 +1,618 @@
+IP addresses and require custom communications implementation to reach the endpoints.
+Such a layer would also enable audit logging of requests made by the API server and allows for more validation of outgoing API server connections.
+It is worth noting that the majority of this traffic is likely to be encrypted within our tunnel. So the level of audit capability added here may be limited.
+Structuring the API server in this way would also be a forcing function for keeping architectural layering violations out of apiserver.
+This separation of networks can be used to protect against security concerns such as [Security Impact of Kubernetes API server external IP address proxying](https://groups.google.com/d/msg/kubernetes-security-announce/tyd-MVR-tY4/tyREP9-qAwAJ).
+
+### Goals
+
+Allow us to delete the SSH Tunnel/Node Dialer code from the Kube API
+Server.
+Provide a mechanism for fixing
+https://groups.google.com/d/msg/kubernetes-security-announce/tyd-MVR-tY4/tyREP9-qAwAJ.
+Allow isolation of the Control network from the Cluster network.
+
+### Non-Goals
+
+Build a general purpose Proxy which does everything.
+The proxy provided can be used as a guide for users to build their own custom
+proxies.
+
+## Definitions
+
+- **Master Network** An IP reachable network space which contains the master components, such as Kubernetes API Server, Connectivity Proxy and ETCD server.
+- **Cluster Network** An IP reachable network space which contains (some/all?) the clusters Nodes. Worth noting that the Cluster Network may be completely disjoint from the Master network, have overlapping IP addresses to the Master Network or other means of network isolation. Direct IP routability between cluster and master networks should not be assumed.
+- **KAS** Kubernetes API Server, responsible for serving the Kubernetes API to clients.
+- **KMS** Key Management Service, plugins for secrets encryption key management
+- **Connectivity service** The component that is built in KAS to provide dialer for outgoing connection request based on NetworkContext information
+- **Connectivity proxy** The proxy that runs in master network, and it has secure channel established to cluster network. It exposes gRPC interface to KAS to provide connectivity service.
+- **Flat Network** A network where traffic can be successfully routed using IP. Implies that there are no overlapping (i.e. shared) IPs on the network.
+
+## Proposal
+
+The proposal is to run connectivity proxy inside master network. 
+The connectivity proxy exposes a gRPC interface to API server. 
+Each connectivity proxy allows secure connections to one or more cluster networks. 
+Any network addressed by a connectivity proxy must be flat. 
+Currently the only mechanism for handling overlapping IP ranges in Kubernetes is the Proxy. 
+As we are passed the proxy there would need to be a non Kubernetes mechanism 
```

Comment:
This was not clear to me.  I think you mean "the Cluster Network must have a mechanism to reach the proxy, and the the KAS must have a mechanism to reach the proxy.  This would be configured as part of cluster creation, not by Kubernetes itself" (or something like that?)